pkg/kubectl/cmd/rollingupdate.go

```diff
@@ -282,21 +291,30 @@ func addDeploymentKeyToReplicationController(oldRc *api.ReplicationController, c
 			return err
 		}
 	}
+
 	if oldRc.Spec.Selector == nil {
 		oldRc.Spec.Selector = map[string]string{}
 	}
-	if oldRc.Spec.Template.Labels == nil {
-		oldRc.Spec.Template.Labels = map[string]string{}
+	// Copy the old selector, so that we can scrub out any orphaned pods
+	selectorCopy := map[string]string{}
+	for k, v := range oldRc.Spec.Selector {
+		selectorCopy[k] = v
 	}
 	oldRc.Spec.Selector[deploymentKey] = oldHash
-	oldRc.Spec.Template.Labels[deploymentKey] = oldHash
 
 	if _, err := client.ReplicationControllers(namespace).Update(oldRc); err != nil {
 		return err
 	}
-	// Note there is still a race here, if a pod was created during the update phase.
-	// It's unlikely, but it could happen, and if it does, we'll create extra pods.
-	// TODO: Clean up orphaned pods here.
+
+	podList, err = client.Pods(namespace).List(labels.SelectorFromSet(selectorCopy), fields.Everything())
```

Comment:
The race is caused by the fact that despite updating the RC in the API server, we have no guarantees that the controller-manager has actually picked up the changed RC, when we subsequently list the existing pods.

The race goes like this:

* We update the RC, but it is not yet picked up by the controller-manager.
* We list the existing pods
* One of those existing pods is deleted
* The controller manager, which hasn't picked up the updated RC, sees the missing Pod and creates a new Pod using the old template.  Because we've already listed the pods, it's not in the list of pods to update
* The controller manager picks up the new RC definition.
* This newly created pod is orphaned.

I agree that it is a pretty unlikely race, but it's definitely possible.